### ISSUE TITLE
Continuing training from checkpoint is different than uninterupted training

### ISSUE THREAD

**Issue by nbrgr** [joeynmt/joeynmt#125]
Currently the checkpoints have no way to continue training from the same point in the dataset that the checkpoint left off at. Instead, the dataset is reset to how it was at the beginning of training using the same state given by the random seed. 

This causes the following training progressions when training for two epochs and interrupting training after the first epoch:

| | Uninterrupted training  | Interrupted training  | Continued from checkpoint |
|---|---|---|---|
| Epoch 1 |    |   |   |
|  | 1 | 1 |  |
|  | 2 | 2 |  |
|  | 3 | 3 |  |
|  | 4 | 4 |  |
| Epoch 2 |  | | |
| | 5 | | 1 |
| | 6 | | 2 |
| | 7 | | 3 |
| | 8 | | 4 |

So the interrupted training plus the continued training causes the model to see the same batches in the same order twice. If the training continued from the checkpoint is allowed to continue for a third epoch, it will then see what the uninterrupted training saw during the second epoch. In order to make sure uninterrupted training is comparable to continuing from checkpoint, it would be good if the checkpoint started and saw batches [5, 6, 7, 8].

Additionally, if the config file specifies two epochs, uninterrupted training will see exactly two epochs, but if you see one epoch before interrupting and continue from that checkpoint, the training will forget that it already saw one epoch and attempt to train on two more epochs instead of one more.

The training iterator also has a state_dict that can be saved and restored, which we should add to the checkpoints.

**Comment by bastings** [comment]
@juliakreutzer I think we can accept a PR for this, do you agree?

**Comment by bastings** [comment]
Also thank you @nbrgr for catching this! :-)

**Comment by may-** [comment]
Hi @nbrgr,

I'm not sure if it is a good idea to save the data iterator's status in a checkpoint. There is no guarantee that the data is exactly the same as before, i.e. the number of minibatch could be different in the resumed training.

**Comment by nbrgr** [comment]
I think there was a comment re: saving the iterator state in the checkpoint. Specifically what would be saved is the state dict

    def state_dict(self):
        return {
            "iterations": self.iterations,
            "iterations_this_epoch": self._iterations_this_epoch,
            "random_state_this_epoch": self._random_state_this_epoch}

which doesn't have any references specifically to the data set. Instead it is the rng state and also keeps track of iterations. So you could alter the data set and it would be as if the data set were dynamically changed during runtime. And there would also be the option to reset the state if you didn't want to use it.

**Comment by may-** [comment]
Hi @nbrgr,

sorry for my mistake. please ignore me!
I wrote my comment before checking your pull request. I like your solution.
